python/ql/src/experimental/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -0,0 +1,312 @@
+ * to the value before the update with the exception of `ObjectCreation`,
+ * which represents the value after the constructor has run.
+ */
+abstract class PostUpdateNode extends Node {
+  /** Gets the node before the state update. */
+  abstract Node getPreUpdateNode();
+}
+
+class DataFlowExpr = Expr;
+
+/**
+ * Flow between ESSA variables.
+ * This includes both local and global variables.
+ * Flow comes from definitions, uses and refinements.
+ */
+// TODO: Consider constraining `nodeFrom` and `nodeTo` to be in the same scope.
+module EssaFlow {
+  predicate essaFlowStep(Node nodeFrom, Node nodeTo) {
+    // Definition
+    //   `x = f(42)`
+    //   nodeFrom is `f(42)`, cfg node
+    //   nodeTo is `x`, essa var
+    nodeFrom.asCfgNode() = nodeTo.asEssaNode().getDefinition().(AssignmentDefinition).getValue()
+    or
+    // Use
+    //   `y = 42`
+    //   `x = f(y)`
+    //   nodeFrom is `y` on first line, essa var
+    //   nodeTo is `y` on second line, cfg node
+    nodeFrom.asEssaNode().getAUse() = nodeTo.asCfgNode()
+    or
+    // Refinements
+    exists(EssaEdgeRefinement r |
+      nodeTo.asEssaNode() = r.getVariable() and
+      nodeFrom.asEssaNode() = r.getInput()
+    )
+    or
+    exists(EssaNodeRefinement r |
+      nodeTo.asEssaNode() = r.getVariable() and
+      nodeFrom.asEssaNode() = r.getInput()
+    )
+    or
+    exists(PhiFunction p |
+      nodeTo.asEssaNode() = p.getVariable() and
+      nodeFrom.asEssaNode() = p.getShortCircuitInput()
+    )
+  }
+}
+
+//--------
+// Local flow
+//--------
+
+/**
+ * This is the local flow predicate that is used as a building block in global
+ * data flow. It is a strict subset of the `localFlowStep` predicate, as it
+ * excludes SSA flow through instance fields.
+ */
+predicate simpleLocalFlowStep(Node nodeFrom, Node nodeTo) {
+  EssaFlow::essaFlowStep(nodeFrom, nodeTo)
+}
+
+// TODO: Make modules for these headings
+//--------
+// Global flow
+//--------
+
+/** Represents a callable */
+class DataFlowCallable = CallableValue;
+
+/** Represents a call to a callable */
+class DataFlowCall extends CallNode {
+  DataFlowCallable callable;
+
+  DataFlowCall() {
+    this = callable.getACall()
+  }
+
+  /** Gets the enclosing callable of this call. */
+  DataFlowCallable getEnclosingCallable() { result = callable }
+}
+
+/** A data flow node that represents a call argument. */
+class ArgumentNode extends Node {
+  ArgumentNode() {
+    exists(DataFlowCall call, int pos |
+      this.asCfgNode() = call.getArg(pos)
+    )
+  }
+
+  /** Holds if this argument occurs at the given position in the given call. */
+  predicate argumentOf(DataFlowCall call, int pos) {
+    this.asCfgNode() = call.getArg(pos)
+  }
+
+  /** Gets the call in which this node is an argument. */
+  final DataFlowCall getCall() { this.argumentOf(result, _) }
+}
+
+/** Gets a viable run-time target for the call `call`. */
+DataFlowCallable viableCallable(DataFlowCall call) {
+  result = call.getEnclosingCallable()
```

Comment:
This looks wrong; `call.getEnclosingCallable()` should give the callable in which `call` occurs, whereas `viableCallable` should give the call graph, i.e. the call-target of a given call.

Reply:
Ah, thanks for the clarification.

Reply:
Should be better now..